src/base/utils/string.h

```diff
@@ -45,8 +45,8 @@ namespace Utils
         // Taken from https://crackstation.net/hashing-security.htm
         bool slowEquals(const QByteArray &a, const QByteArray &b);
 
-        bool naturalCompareCaseSensitive(const QString &left, const QString &right);
```

Comment:
Can we have only one function naturalCompare with 3rd parameter indicating case sensitivity? (of course if change NaturalCompare class to have case sensitivity param in operator() instead of constructor, or even get rid from this class unless it used somewhere else).

Reply:
I don't remember how it started... Why are you using a static (thread local) NaturalCompare object? It seems the design would be simpler without this approach. BTW, application won't exit correctly due to NaturalCompare if built with MinGW on Windows.

Reply:
>even get rid from this class unless it used somewhere else

this function is still used in 1 place: gui/transferlistfilterswidget.cpp

>Why are you using a static (thread local) NaturalCompare object? 

I think it's for thread safety...

Reply:
>I think it's for thread safety...

No, I mean why are you using thread local (i.e. thread wide static) object instead of creating it inplace?

Reply:
>No, I mean why are you using thread local (i.e. thread wide static) object instead of creating it inplace?

I speculate creating QCollator (in NaturalCompare) over and over in a tight loop (std::sort) could have a performance hit.

Reply:
>Can we have only one function naturalCompare with 3rd parameter indicating case sensitivity? 

done.

Reply:
The doc says:
>A QCollator object can be used together with template based sorting algorithms such as std::sort to sort a list of QStrings.

But it doesn't say "this can be slow".
You only need to prevent its re-creation. You can create object before perform sort, then capture it in used lambda, or even implement NaturalLessThan class and use it in std::sort.

Reply:
Right, but what is the problem with current implementation?  it already handles all these issues, although ugly.

Reply:
>BTW, application won't exit correctly due to NaturalCompare if built with MinGW on Windows.

Reply:
>application won't exit correctly due to NaturalCompare if built with MinGW on Windows.

Is it caused by this PR?
I don't build with mingw, are there more information you can provide?

Reply:
No, this is issue of current code.
My main build environment is MSVC-based, so I deal with MinGW quite rare. Each time I run qBittorrent under MinGW environment I have noticed it won't be exited successfully. Once I tried to debug it and I found it crashes due to NaturalCompare destruction. I had no time to debug it deeply.

Reply:
@glassez:  MinGW binaries, as a general rule, have problems with threading on Windows.  

I found that out a few weeks ago, while I was attempting to research the performance differences between binaries built by MSVS and ones built by MinGW…

Reply:
Really, I have not found any bugs in our code causing this crash and I also think it's problem of MinGW. So if you're not going to get rid of the "ugliness", I will not insist on proposed changes. Personally I would implement NaturalLessThan functor and use it with std::sort etc.

Reply:
>So if you're not going to get rid of the "ugliness", I will not insist on proposed changes. 

I'll leave it be.

>Personally I would implement NaturalLessThan functor and use it with std::sort etc.

Make sense, added a helper function.